universal-application-tool-0.0.1/app/repository/VersionRepository.java

```diff
@@ -0,0 +1,328 @@
+      // For each unedited active program which does not have a draft program of the
+      // same name, set the version and go.
+      long version = getNextVersion();
+      for (Program uneditedActive :
+          ebeanServer
+              .find(Program.class)
+              .alias("active")
+              .where()
+              .eq("lifecycle_stage", LifecycleStage.ACTIVE)
+              .notExists(
+                  ebeanServer
+                      .find(Program.class)
+                      .where()
+                      .raw("active.name = name")
+                      .eq("lifecycle_stage", LifecycleStage.DRAFT)
+                      .query())
+              .findList()) {
+        uneditedActive.setLifecycleStage(LifecycleStage.OBSOLETE);
+        uneditedActive.save();
+        Program newVersion = new Program(uneditedActive.getProgramDefinition());
+        newVersion.setVersion(version);
+        newVersion.setLifecycleStage(LifecycleStage.ACTIVE);
+        this.programRepository.insertProgramSync(newVersion);
+      }
+
+      // Then, for each program, if a draft exists, publish the draft.
+      ebeanServer
+          .find(Program.class)
+          .where()
+          .eq("lifecycle_stage", LifecycleStage.DRAFT)
+          .findList()
+          .stream()
+          .forEach(
+              draft -> {
+                draft.setVersion(version);
+                this.publishProgram(draft, transaction);
+              });
+      ebeanServer.commitTransaction();
+    } finally {
+      ebeanServer.endTransaction();
+    }
+  }
+
+  public long getNextVersion() {
+    Optional<Program> draft =
+        ebeanServer
+            .find(Program.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.DRAFT)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (draft.isPresent()) {
+      return draft.get().getVersion();
+    }
+    // If there are no drafts, add one to any active program
+    Optional<Program> active =
+        ebeanServer
+            .find(Program.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.ACTIVE)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (active.isPresent()) {
+      return active.get().getVersion() + 1;
+    }
+
+    // If there are no programs, we are pretty sure the correct version is 1, but
+    // we might also be in a questions-only test.  Search questions just in case.
+    Optional<Question> draftQuestion =
+        ebeanServer
+            .find(Question.class)
+            .where()
+            .eq("lifecycle_stage", LifecycleStage.DRAFT)
+            .isNotNull("version")
+            .setMaxRows(1)
+            .findOneOrEmpty();
+    if (draftQuestion.isPresent()) {
+      return draftQuestion.get().getVersion();
+    }
+    // If there are no drafts, add one to any active program
```

Comment:
Suggestion (blocking): 
```suggestion
    // If there are no drafts, add one to any active question
```

Reply:
Should this be adding one to any active program? It seems like `Program` has precedence over `Question` in the first half of this code, which doesn't match basing the next version number on an active `Question` and not an active `Program` first.

Reply:
See Kendrick's question

Reply:
This part of the code is only in here so that we can correctly calculate versions in the specifically constructed case where we have inserted questions in the active state during testing.  You're right that programs have precedence, this comment is below the part where we have checked on programs.  We are basing this on a question first.